### PR TITLE
Declarative DSL

### PR DESCRIPTION
**Pull request by novemberborn**
Implements most of #33. Supports `app-action`, `app-actions` and `app-store`. See <https://github.com/novemberborn/app-example/commit/18d48def9cc7f5640bcbf7e11566b357cff30e84> for an (outdated) example.

Now also allows initial state to be defined in action definitions (and declaratively). Reverts #21 for consistency. Action factories now receive a single options object, with a `RegistryProvider` instead of a `CombinedRegistry`. The `RegistryProvider` is also used when configuring actions.

Remaining work:

* [x] Declarative widget and custom element registrations
* [x] Documentation

### CODE REVIEW

**Comment by novemberborn** [comment]
This in now code complete. Just need to finish adding documentation to the README. I encourage all to go through the commits, especially as I had to do some refactoring and add some functionality to the functional API.

**Comment by novemberborn** [comment]
@kitsonk noticed how `data-options` for custom elements allows you to specify `id`, `listeners` and `stateFrom`. That's weird because we don't allow those for any other definition. Pushed a commit which removes that support.

**Comment by novemberborn** [comment]
Now with documentation!

**Comment by novemberborn** [comment]
FWIW this should be merged manually, lots of atomic commits that should be preserved.

**Comment by novemberborn** [comment]
Rebased.

@kitsonk et al, feedback welcome 😄

File: src/lib/moduleResolver.ts [hunk 1]
```diff
@@ -14,18 +15,29 @@ export interface ToAbsMid {
...
 		return new Promise((resolve) => {
 			// Assumes require() is an AMD loader!
-			require([toAbsMid(mid)], (module) => {
-				if (module.__esModule) {
-					resolve(module.default);
-				}
-				else {
-					resolve(module);
+			require([toAbsMid(mid)], resolve);
+		}).then((module: any) => {
```
Comment (kitsonk) [inline src/lib/moduleResolver.ts]:
```
})
.then((
```

**Comment by novemberborn** [comment]
Splitting into two new PRs.

**Pull request closed without merging**